### PR TITLE
Listen for Zookeeper change events to trigger refresh.

### PR DESCRIPTION
**Pull request by spencergibb**
fixes gh-49

As it was implemented, `ZookeeperTreeCachePropertySource` was broken if `/refresh` was called (since `close()` was called it remained empty). This decouples watching of the tree from property sources and doesn't involve any acrobatics to inject an `ApplicationEventPublisher` into a property source.

I believe this satisfies the general idea of the `ZookeeperTreeCachePropertySource`.

This would supersede #55 by @efenderbosch

/cc @marcingrzejszczak @dsyer

### CODE REVIEW

File: spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher publisher) {
+		this.publisher = publisher;
+	}
+
+	@PostConstruct
+	public void start() {
+		if (this.running.compareAndSet(false, true)) {
+			caches = new HashMap<>();
```
Comment (marcingrzejszczak) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java]: according to https://github.com/spring-cloud/spring-cloud-build/blob/master/spring-cloud-build-tools/src/main/resources/checkstyle.xml there should be `this.` in front of every field

File: spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public class ZookeeperConfigAutoConfiguration {
+
+	@Configuration
+	@ConditionalOnClass(RefreshEndpoint.class)
+	protected static class ZkRefreshConfiguration {
+		@Bean
+		@ConditionalOnBean(RefreshEndpoint.class)
+		public ZookeeperConfigRefreshListener zookeeperConfigRefreshListener(
+				RefreshEndpoint refreshEndpoint) {
+			return new ZookeeperConfigRefreshListener("application", refreshEndpoint);
```
Comment (marcingrzejszczak) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigAutoConfiguration.java]: Why `application` is fixed? Will it always have that value?

Comment (spencergibb) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigAutoConfiguration.java]: It's leftover from an experiment, will remove.

File: spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigRefreshListener.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	@EventListener
+	public void handle(ApplicationReadyEvent event) {
+		this.ready.compareAndSet(false, true);
+	}
+
+	@EventListener
+	public void handle(ZookeeperConfigRefreshEvent event) {
+		if (this.ready.get()) { // don't handle events before app is ready
+			log.debug(contextName + " " + event.getEventDesc());
+			if (this.refresh != null) {
```
Comment (marcingrzejszczak) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigRefreshListener.java]: If there is no `RefreshEndpoint` the bean won't be instantiated, so I guess the null check is unnecessary?

```
@Bean
 		@ConditionalOnBean(RefreshEndpoint.class)
 		public ZookeeperConfigRefreshListener zookeeperConfigRefreshListener(
 				RefreshEndpoint refreshEndpoint) {
 			return new ZookeeperConfigRefreshListener("application", refreshEndpoint);
}
```

File: spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java (hunk 1 above)
Comment (spencergibb) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java]: does checkstyle run?

Comment (marcingrzejszczak) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java]: I guess you have to add it manually like here - https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/pom.xml#L95

**Comment by marcingrzejszczak** [comment]
The implementation looks ok - will you write some tests to back it up?

File: spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java (hunk 1 above)
Comment (marcingrzejszczak) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java]: BTW I've added this https://github.com/spring-cloud/spring-cloud-build/blob/master/spring-cloud-build-tools/src/main/resources/checkstyle.xml to InteliJ checkstyle plugin and I've got info at runtime that sth's wrong

Comment (spencergibb) [inline spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ConfigWatcher.java]: nice

**Comment by spencergibb** [comment]
Squashed and closed via 810fdad74995b971ca1242072ad08ff1a03c92c6

**Pull request closed without merging**